### PR TITLE
Improved mega error handling

### PR DESCRIPTION
**Pull request by naz**
@allouis @kevinansfield it's a WIP but something you guys could pick up and continue (e.g. introduce separate field in emails table to save emails of failed recipients)

### CODE REVIEW

**Comment by gargol** [comment]
Some observations regarding errors in Mailgun. First of all, they are objectively :hankey: as well as lack of any mentions in the docs. 
- Confirmed that 1000 is a hard limit for a batch (received a 400 error when making a bigger batch)
- When sending emails through sandbox and batch consists of validated and unvalidated emails the whole batch fails. In a way it's good because we can retry the whole batch, on the other had it's bad because one bad apple could spoil the batch :/  
- We need some more reliable structure to store failed batches. Right now the table becomes bloated if the batch fails and it stores at least this many characters:
```
// (M) max email length
// (B) max batch size = 1000
// (
// 191 x 1000 = 191000 (minimum of what we need to store)
// plus error details and other json fluf that goes along with it
```
- If more then one batch fails above multiplies!
- There's no way to reliably predict what kinds of errors will happen (no docs yey!), so taking a brute-force approach with storing as much as possible now and we should monitor the behavior closely for refinements.

File: core/server/services/bulk-email/index.js [hunk 1]
```diff
@@ -36,33 +36,56 @@ module.exports = {
...
-                    Object.assign(messageData, {
-                        'o:tag': [bulkEmailConfig.mailgun.tag, 'bulk-email']
-                    });
-                }
+            return new Promise((resolve) => {
+                mailgunInstance.messages().send(messageData, (error, body) => {
+                    if (error) {
+                        // NOTE: logging an error here only but actual handling should happen in more sophisticated batch retry handler
+                        // REF: possible mailgun errors https://documentation.mailgun.com/en/latest/api-intro.html#errors
+                        common.logging.error(new common.errors.GhostError({
```
Comment (allouis) [inline core/server/services/bulk-email/index.js]: Should probably be a `warn` maybe?

File: core/server/services/bulk-email/index.js [hunk 2]
```diff
@@ -36,33 +36,56 @@ module.exports = {
...
+                            err: error,
+                            context: 'The bulk email service was unable to send a message, your site will continue to function.',
+                            help: common.i18n.t('errors.services.ping.requestFailed.help', {url: 'https://ghost.org/docs/'})
+                        }));
 
-                return mailgunInstance.messages().send(messageData);
+                        // NOTE: these are generated variables, so can be regenerated when retry is done
+                        const data = _.omit(batchData, ['recipient-variables']);
+                        resolve({
```
Comment (allouis) [inline core/server/services/bulk-email/index.js]: Maybe we wanna resolve with something like:
```js
resolve(Success(data))
// OR
resolve(Failure(data))
```

Check out the Either monad for inspiration :stuck_out_tongue:

**Comment by kevinansfield** [comment]
> When sending emails through sandbox and batch consists of validated and unvalidated emails the whole batch fails. In a way it's good because we can retry the whole batch, on the other had it's bad because one bad apple could spoil the batch :/

Do we get details on which emails failed? If so we will probably want some way to filter those out on automated retries in the future.

At some point we probably also want to incorporate Mailgun's validation checks on emails https://documentation.mailgun.com/en/latest/user_manual.html#email-validation-v4. We want to do as much as possible to preserve our sender reputation.

**Comment by gargol** [comment]
@kevinansfield 
> Do we get details on which emails failed? If so we will probably want some way to filter those out on automated retries in the future.

No details, just a whole batch failure.

> At some point we probably also want to incorporate Mailgun's validation checks on emails https://documentation.mailgun.com/en/latest/user_manual.html#email-validation-v4. We want to do as much as possible to preserve our sender reputation.

This would definitely be good. But I think we don't need to worry about it much at the moment because for most cases members come into the db through email validation, so it's rather edging case for somebody to be completely invalid (should be manually created or something).

File: core/server/translations/en.json [hunk 3]
```diff
@@ -519,6 +519,11 @@
                 },
                 "loader": "Error trying to load YAML setting for {setting} from '{path}'.",
                 "ensureSettings": "Error trying to access settings files in {path}."
+            },
+            "mega": {
+                "requestFailed": {
+                    "error" : "The email service was unable to send an email batch."
```
Comment (gargol) [inline core/server/translations/en.json]: @peterzimon would love a wording check from you. This is used when the batch is failing.

**Pull request merged**